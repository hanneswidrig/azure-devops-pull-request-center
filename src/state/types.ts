import * as DevOps from 'azure-devops-extension-sdk';
import { ObservableValue } from 'azure-devops-ui/Core/Observable';
import { IdentityRef } from 'azure-devops-extension-api/WebApi/WebApi';
import { WorkItem } from 'azure-devops-extension-api/WorkItemTracking/WorkItemTracking';
import { IdentityRefWithVote, PullRequestStatus, GitRepository } from 'azure-devops-extension-api/Git/Git';

import { TabOptions } from '../tabs/Tabs.types';
import { ReviewerVoteLabel } from '../lib/enums';

const SET_REPOSITORIES = 'setRepositories';
const SET_PULL_REQUESTS = 'setPullRequests';
const SET_CURRENT_USER = 'setCurrentUser';
const SET_SELECTED_TAB = 'setSelectedTab';
const ADD_ASYNC_TASK = 'addAsyncTask';
const REMOVE_ASYNC_TASK = 'removeAsyncTask';
const TOGGLE_FILTER_BAR = 'toggleFilterBar';
const DISPLAY_WORK_ITEMS = 'displayWorkItems';
const REFRESH_PULL_REQUESTS = 'refreshPullRequests';
const TOGGLE_FULL_SCREEN_MODE = 'toggleFullScreenMode';

export const ActionTypes = {
  SET_REPOSITORIES,
  SET_PULL_REQUESTS,
  SET_CURRENT_USER,
  SET_SELECTED_TAB,
  ADD_ASYNC_TASK,
  REMOVE_ASYNC_TASK,
  TOGGLE_FILTER_BAR,
  DISPLAY_WORK_ITEMS,
  REFRESH_PULL_REQUESTS,
  TOGGLE_FULL_SCREEN_MODE
} as const;

export interface PR {
  pullRequestId: number;
  repositoryId: string;
  isDraft: boolean;
  isAutoComplete: boolean;
  status: PullRequestStatus;

  title: string;
  href: string;
  createdBy: IdentityRef;
  creationDate: Date;
  secondaryTitle: string;

  sourceBranch: PRRef;
  targetBranch: PRRef;
  repository: PRRef;

  myApprovalStatus: ReviewerVoteLabel;

  workItems: WorkItem[];
  reviewers: IdentityRefWithVote[];
}

type PRRef = { name: string; href: string };

export interface Data {
  repositories: GitRepository[];
  pullRequests: PR[];
  currentUser: DevOps.IUserContext;
  asyncTaskCount: number;
}

export interface UI {
  isFilterVisible: ObservableValue<boolean>;
  isFullScreenMode: boolean;
  selectedTab: TabOptions;
}

export interface PrHubState {
  data: Data;
  ui: UI;
}
