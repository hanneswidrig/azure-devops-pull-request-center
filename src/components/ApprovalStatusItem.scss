.asi-container {
  padding: 5px 0;
  display: flex;
}
