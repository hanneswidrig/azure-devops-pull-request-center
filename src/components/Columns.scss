@import "azure-devops-ui/Core/_platformCommon.scss";

.work-button > span:last-child {
  font-size: 12px;
  line-height: 20px;
}

td.bolt-table-cell-content.no-padding.bolt-table-cell.bolt-list-cell > div {
  justify-content: center;
}

.invisible {
  opacity: 0;
  position: absolute;
}

.tooltip-overflow-parent {
  background-color: $backgroundColor;
  color: $backgroundColor;
  padding: 16px;
  border-radius: 4px;
}

.tooltip-overflow-child + .tooltip-overflow-child {
  margin-top: 8px;
}

.tooltip-overflow .bolt-tooltip-content {
  margin: 0;
  padding: 0;
  background-color: transparent;
  box-shadow: 0 1.6px 3.6px 0 rgba(0, 0, 0, 0.132), 0 0.3px 0.9px 0 rgba(0, 0, 0, 0.108);
  box-shadow: 0 1.6px 3.6px 0 var(--callout-shadow-color, rgba(0, 0, 0, 0.132)),
    0 0.3px 0.9px 0 var(--callout-shadow-secondary-color, rgba(0, 0, 0, 0.108));
}
