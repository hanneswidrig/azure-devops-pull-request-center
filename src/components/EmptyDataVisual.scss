@import 'node_modules/azure-devops-ui/Core/_platformCommon.scss';

.empty-data-container {
  width: 100%;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  background-color: $backgroundColor;
  color: $primary-text;
}

.empty-data-text {
  font-size: 1.25rem;
  text-align: center;
  margin: 4rem 0;
}
