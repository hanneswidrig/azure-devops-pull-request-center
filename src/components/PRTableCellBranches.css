button.branch-button {
  height: 32px;
  overflow: hidden;
}

button.branch-button:hover,
button.branch-button:active {
  color: var(--palette-primary-darkened-10);
}

.branch-button > span:last-child {
  font-size: 12px;
  line-height: 16px;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  max-width: 256px;
}
