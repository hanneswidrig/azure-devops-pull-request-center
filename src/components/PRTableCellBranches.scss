@import '~azure-devops-ui/Core/platformCommon';

button.branch-button {
  height: 32px;
  overflow: hidden;
}

button.branch-button:hover,
button.branch-button:active {
  color: $communication-background-active;
}

.branch-button > span:last-child {
  font-size: 12px;
  line-height: 16px;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  max-width: 256px;
}
