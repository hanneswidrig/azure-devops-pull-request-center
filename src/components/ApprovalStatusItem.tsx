import React from 'react';

import { ITableColumn } from 'azure-devops-ui/Table';
import { IListBoxItem } from 'azure-devops-ui/ListBox';

import './ApprovalStatusItem.scss';
import { getReviewerVoteIconStatus } from './StatusIcon';

export const ApprovalStatusItem = (
  rowIndex: number,
  columnIndex: number,
  tableColumn: ITableColumn<IListBoxItem<{}>>,
  tableItem: IListBoxItem<{}>,
) => {
  return (
    <td key={rowIndex} className="bolt-list-box-text bolt-list-box-text-multi-select asi-container">
      {getReviewerVoteIconStatus(tableItem.id)}
      <span className="margin-left-8">{tableItem.text}</span>
    </td>
  );
};
