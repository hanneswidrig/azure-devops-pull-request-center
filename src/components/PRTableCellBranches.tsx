import React from 'react';
import { Button } from 'azure-devops-ui/Button';

import './PRTableCellBranches.scss';
import { PR } from '../state/types';

export const PRTableCellBranches: React.FC<{ tableItem: PR }> = ({ tableItem }: { tableItem: PR }) => {
  const sourceBranch = tableItem.sourceBranch.name.replace('refs/heads/', '');
  const targetBranch = tableItem.targetBranch.name.replace('refs/heads/', '');

  return (
    <div className='flex-row flex-center rhythm-horizontal-4'>
      <Button
        className='branch-button'
        text={tableItem.repository.name}
        iconProps={{ iconName: 'Repo' }}
        onClick={() => {
          navigator.clipboard.writeText(tableItem.repository.name);
        }}
        subtle
      />
      <Button
        className='branch-button text-ellipsis'
        text={sourceBranch}
        iconProps={{ iconName: 'BranchCompare' }}
        tooltipProps={{ text: sourceBranch, delayMs: 500 }}
        onClick={() => {
          navigator.clipboard.writeText(sourceBranch);
        }}
        subtle
      />
      <Button
        className='branch-button'
        text={targetBranch}
        iconProps={{ iconName: 'BranchMerge' }}
        onClick={() => {
          navigator.clipboard.writeText(targetBranch);
        }}
        subtle
      />
    </div>
  );
};
