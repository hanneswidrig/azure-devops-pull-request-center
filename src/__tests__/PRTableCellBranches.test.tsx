import React from 'react';
import '@testing-library/jest-dom/extend-expect';
import { render, cleanup } from '@testing-library/react';
import { PRTableCellBranches } from '../components/PRTableCellBranches';
import { PR } from '../state/types';

afterEach(cleanup);

describe('<PRTableCellBranches />', () => {
  test('render with correct text for each button', () => {
    const props = {
      repository: {
        name: 'Repository',
      },
      sourceBranch: {
        name: 'refs/heads/Source Branch',
      },
      targetBranch: {
        name: 'refs/heads/Target Branch',
      },
    } as PR;
    const { container, getByText } = render(<PRTableCellBranches tableItem={props} />);
    expect(container).toMatchSnapshot();
    expect(getByText('Repository')).toBeTruthy();
    expect(getByText('Source Branch', { exact: true })).toBeTruthy();
    expect(getByText('Target Branch', { exact: true })).toBeTruthy();
  });
});
