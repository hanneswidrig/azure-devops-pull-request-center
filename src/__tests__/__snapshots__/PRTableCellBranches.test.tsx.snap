// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PRTableCellBranches /> render with correct text for each button 1`] = `
<div>
  <div
    class="flex-row flex-center rhythm-horizontal-4"
  >
    <button
      class="branch-button bolt-button bolt-icon-button enabled subtle bolt-focus-treatment"
      data-focuszone=""
      data-is-focusable="true"
      role="button"
      tabindex="0"
      type="button"
    >
      <span
        aria-hidden="true"
        class="left-icon flex-noshrink fabric-icon ms-Icon--Repo medium"
      />
      <span
        class="bolt-button-text body-m"
      >
        Repository
      </span>
    </button>
    <button
      class="branch-button text-ellipsis bolt-button bolt-icon-button enabled subtle bolt-focus-treatment"
      data-focuszone=""
      data-is-focusable="true"
      role="button"
      tabindex="0"
      type="button"
    >
      <span
        aria-hidden="true"
        class="left-icon flex-noshrink fabric-icon ms-Icon--BranchCompare medium"
      />
      <span
        class="bolt-button-text body-m"
      >
        Source Branch
      </span>
    </button>
    <button
      class="branch-button bolt-button bolt-icon-button enabled subtle bolt-focus-treatment"
      data-focuszone=""
      data-is-focusable="true"
      role="button"
      tabindex="0"
      type="button"
    >
      <span
        aria-hidden="true"
        class="left-icon flex-noshrink fabric-icon ms-Icon--BranchMerge medium"
      />
      <span
        class="bolt-button-text body-m"
      >
        Target Branch
      </span>
    </button>
  </div>
</div>
`;
